source/common/http/async_client_impl.h

```diff
@@ -93,7 +98,22 @@ class AsyncStreamImpl : public AsyncClient::Stream,
 public:
   AsyncStreamImpl(AsyncClientImpl& parent, AsyncClient::StreamCallbacks& callbacks,
                   const AsyncClient::StreamOptions& options);
-  ~AsyncStreamImpl() override { router_.onDestroy(); }
+  ~AsyncStreamImpl() override {
+    router_.onDestroy();
+    if (destructor_callback_.has_value()) {
+      (*destructor_callback_)();
+    }
+  }
+
+  void setDestructorCallback(AsyncClient::StreamDestructorCallbacks callback) override {
+    destructor_callback_.emplace(callback);
+  }
+  void removeDestructorCallback() override { destructor_callback_.reset(); }
+
+  void setWatermarkCallbacks(DecoderFilterWatermarkCallbacks& callbacks) override {
+    watermark_callbacks_.emplace(callbacks);
+  }
+  void removeWatermarkCallbacks() override { watermark_callbacks_.reset(); }
```

Comment:
ditto here - if we have a shadow request from cluster A to cluster B, cluster B is slow and causes back-up so we pause reading from the client, shadow request is reset remotely, so we kill off that client, will we resume reading from the client and spooling to A?

Reply:
Done.